department-of-veterans-affairs/caseflow
# Lianna/APPEALS-13211: Handle team management mailing functionality

Resolves [APPEALS-13211](https://vajira.max.gov/browse/APPEALS-13211)

### Description
We need to create a unique class to handle the various email scenarios that we will need to generate.
The passing of the information for each mailing scenario will be handled within the respective stories handling the overall implementation for each.

### Acceptance Criteria
- [x] Creation of a new mailer class to handle the selection of the respective email templates
- [x] Creation of job(s) to handle respective email notification scenarios including:
  - [x] Notifying requestor upon successful membership request submission
  - [x] Notifying requestor upon change of status of membership request
  - [x] Notifying admins upon successful membership request submission

### Testing Plan
1. Go to Terminal and setup Local Email Server:
  1a) `docker run -p 1080:1080 -p 1025:1025 maildev/maildev` Keep this terminal open in it's own tab.
  1b) `WITH_TEST_EMAIL_SERVER=true make run-low-priority` Keep this terminal open in it's own tab.

2. In your web browser navigate to [http://localhost:1080/](http://localhost:1080/ )

<img width="838" alt="Local Email Server" src="https://user-images.githubusercontent.com/17394371/215857356-075d1e6f-884b-4120-a422-eb6ec566ad62.png">

3. Return to Terminal and use the following command to open the rails console: 
`WITH_TEST_EMAIL_SERVER=true make c`
  
4) Set the recipient_info by typing the following into the terminal: 

`recipient_info = { email: "test@test.com" }`

5. When a user submits a membership request they should receive an email confirming the request was submitted successfully. Now run the following commands to send this first email:
  5a) `email_type = "SendMembershipRequestSubmittedEmail"`
  5b) `Memberships::SendMembershipRequestMailerJob.perform_now(email_type, recipient_info)`
  5c) `Memberships::SendMembershipRequestMailerJob.perform_later(email_type, recipient_info)`  

<img width="839" alt="submitted" src="https://user-images.githubusercontent.com/17394371/219500828-96c56814-29b7-49f6-ba81-a21550b25f29.png">

6. Then navigate to the browser and there should be a new email in the Local Email Server: 

<img width="1680" alt="submitted" src="https://user-images.githubusercontent.com/17394371/219247635-63fab46a-21af-4300-8255-0f998dab6e37.png">

7. Repeat Steps 5 & 6 for the next two emails.

8. When an admin updates the status of a membership request the requestor should receive an email with a status update. In Terminal run: 
  8a) `email_type = "SendUpdatedMembershipRequestStatusEmail"`
  8b) `Memberships::SendMembershipRequestMailerJob.perform_later(email_type, recipient_info)`
  8c) `Memberships::SendMembershipRequestMailerJob.perform_now(email_type, recipient_info)`  

<img width="1678" alt="send_admin" src="https://user-images.githubusercontent.com/17394371/219501263-7eff20fa-7229-4f99-bb5b-fcc07ea0ef30.png">

9. In the Browser there should be a new email:

<img width="1680" alt="updated" src="https://user-images.githubusercontent.com/17394371/219247697-c3be92e6-58ed-40ba-9c11-8016268fc4aa.png">

10. When a user submits a request an email notification to admins should be sent. In Terminal run:
  10a) `email_type = "SendAdminsMembershipRequestSubmissionEmail"`
  10b) `Memberships::SendMembershipRequestMailerJob.perform_later(email_type, recipient_info)` 
  10c) `Memberships::SendMembershipRequestMailerJob.perform_now(email_type, recipient_info)` 

<img width="1675" alt="updated_status" src="https://user-images.githubusercontent.com/17394371/219501409-5829fa13-4588-4d26-a77b-c1449ad71bca.png">

11. In Browser the last email should appear:

<img width="1680" alt="submission" src="https://user-images.githubusercontent.com/17394371/219247792-09808a6f-3f5d-4a18-abe0-ab1886622943.png">

12. In a new terminal tab outside of the rails console let's run the testing suites:
`make one-test ./spec/mailers/membership_request_mailer_spec.rb`

<img width="1676" alt="APPEALS_13211_Testing" src="https://user-images.githubusercontent.com/17394371/215856446-e2b26939-714c-44fd-bd92-22a475e6a22d.png">

13. Lastly in terminal run the second testing suite:
`make one-test ./spec/jobs/memberships/send_membership_request_mailer_job_spec.rb`

<img width="1639" alt="send_membership_request_job_spec" src="https://user-images.githubusercontent.com/17394371/219248238-b7e599d6-2f54-4d5f-8d29-e4f3ec0ed77a.png">
